gnoland/website/pkgs/doc/doc.go

```diff
@@ -0,0 +1,96 @@
+package doc
+
+import (
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"sort"
+	"strings"
+)
+
+func New(pkgPath string, files map[string]string) (*Package, error) {
+	p := Package{
+		ImportPath: pkgPath,
+		Filenames:  make([]string, 0, len(files)),
+	}
+
+	fset := token.NewFileSet()
+	gnoFiles := make(map[string]*ast.File)
+
+	for filename, fileContent := range files {
+		p.Filenames = append(p.Filenames, filename)
+
+		f, err := parser.ParseFile(fset, filename, fileContent, parser.ParseComments)
+		if err != nil {
+			return nil, err
+		}
+
+		ast.FileExports(f)
+
+		if strings.HasSuffix(filename, "_test.gno") || strings.HasSuffix(filename, "_filetest.gno") {
+			continue
+		}
+
+		gnoFiles[filename] = f
+
+		if p.Name == "" {
+			p.Name = f.Name.Name
+		}
+
+		if f.Doc != nil {
+			doc := f.Doc.Text()
+			if p.Doc != "" {
+				p.Doc += "\n"
+			}
+			p.Doc += doc
+		}
+	}
+
+	for _, f := range gnoFiles {
+		for _, decl := range f.Decls {
+			switch x := decl.(type) {
+			case *ast.FuncDecl:
+				if x.Name.IsExported() {
```

Comment:
Do you think that public functions of a non-exported type should not be displayed? Because if the type is returned by a constructor, it is useful to have the functions visible in the doc so maybe make this check and in this case display them?

Reply:
I don't think so. I think in current Gno code, there are cases where unexported types are returned by a constructor; but this is not really advisable (if the type is useful to external packages/realms, it should be exported; if we want to keep the fields not accessible, we should use an interface). So while it is allowed by the G(n)o grammar, I've generally never seen it done in idiomatic Go code, such as the standard library. Golint and revive also advise against it.

`gnodev doc` (#522) would still allow inspecting unexported types, so as far as tooling goes I think there will still be ways to consult the documentation for unexported types, if strictly necessary.